gems/sorbet-runtime/tools/generate_call_validation.cc

```diff
@@ -220,6 +197,108 @@ void generateCreateValidatorFast(const Options &options, ValidatorKind kind, siz
+        }
+
+        fmt::print("      create_validator_medium{}(mod, original_method, method_sig, return_type", arity);
+        for (size_t i = 0; i < arity; i++) {
+            fmt::print(",\n");
+            fmt::print("                                    method_sig.arg_types[{}][1]", i);
+        }
+        fmt::print(")\n");
+    }
+    fmt::print("    else\n"
+               "      raise 'should not happen'\n"
+               "    end\n"
+               "  end\n"
+               "\n");
+}
+
+void generateCreateValidatorMedium(const Options &options, size_t arity) {
+    fmt::print("  def self.create_validator_medium{}(mod, original_method, method_sig, return_type", arity);
+    for (size_t i = 0; i < arity; i++) {
+        fmt::print(", arg{}_type", i);
+    }
+    fmt::print(")\n");
+
+    fmt::print("    mod.send(:define_method, method_sig.method_name) do |");
+    for (size_t i = 0; i < arity; i++) {
+        fmt::print("arg{}, ", i);
+    }
+    fmt::print("&blk|\n");
+
+    fmt::print("      # This method is a manually sped-up version of more general code in `validate_call`\n");
+
+    for (size_t i = 0; i < arity; i++) {
+        fmt::print("      if (err = arg{0}_type.error_message_for_obj(arg{0}))\n"
+                   "        CallValidation.report_error(\n"
+                   "          method_sig,\n"
+                   "          err,\n"
+                   "          'Parameter',\n"
+                   "          method_sig.arg_types[{0}][0],\n"
+                   "          arg{0}_type,\n"
+                   "          arg{0},\n"
+                   "          caller_offset: -1\n"
+                   "        )\n"
+                   "      end\n"
+                   "\n",
+                   i);
+    }
+
+    fmt::print("      # The following line breaks are intentional to show nice pry message\n");
+    for (size_t i = 0; i < 10; i++) {
+        fmt::print("\n");
+    }
+    fmt::print("      # PRY note:\n"
+               "      # this code is sig validation code.\n"
+               "      # Please issue `finish` to step out of it\n"
+               "\n");
+
+    fmt::print("      return_value = original_method");
+    if (options.bindCall) {
+        fmt::print(".bind_call(self, ");
+    } else {
+        fmt::print(".bind(self).call(");
+    }
+    for (size_t i = 0; i < arity; i++) {
+        fmt::print("arg{}, ", i);
+    }
+    fmt::print("&blk)\n");
+
+    fmt::print("      if return_type\n"
+               "        if (message = return_type.error_message_for_obj(return_value))\n"
+               "          CallValidation.report_error(\n"
+               "            method_sig,\n"
+               "            message,\n"
+               "            'Return value',\n"
+               "            nil,\n"
+               "            return_type,\n"
+               "            return_value,\n"
+               "            caller_offset: -1\n"
+               "          )\n"
+               "        end\n"
```

Comment:
The other two (methods, procedures) take almost the same codepath but then branch on `validatorKind` in a few places.

It seems like the "medium" path is pretty similar to the fast path, differing only by

- the names of the individual methods
- the condition used to check the types

Do you think it's worth trying to share more code between the different kinds of validators?

Or at least factor out some of the shared code?

Reply:
The handling of `return_type` is different too, but that doesn't affect your point much.

I agree that more code could be shared; I started off with a separate method because I wasn't sure how much would end up being shared, but it ended up being most of it.

I think merging might make things less readable - I already find the switches on ValidatorKind a bit hard to read -but it would make it less likely that some difference could be introduced accidentally, so maybe it's worth it. WDYT?

Reply:
i tend to think that in terms of "hard to read," if you're trying to change this code you're going to be looking at the generated code, imagining how you'd like the generated code to change, and back-solving for a diff to the the generator (i.e., the C++ code is not read in a vacuum, it is read in tandem with the generated code).

Given that, I place more value on the second part (introducing an accidental divergence).

So yeah, if you don't feel strongly one way or another i'd love to share more of the repetitive code between these two.

> but it ended up being most of it

for what it's worth, this is basically the same process and realization that I had when I wrote it initially (it started out as two, and then I realized that so little was different)